Adds a Modern Mode handling for merge requests.

pbendersky opened the pull request:
In Modern Mode, the plugin will fetch Merge Requests from origin instead of using parameterized builds.

This supersedes #252 and works with the refactored 1.2 codebase.

pbendersky committed:
Adds a Modern Mode handling for merge requests. In Modern Mode, the plugin will fetch Merge Requests from origin instead of using parameterized builds.

pbendersky committed:
Fixes for compatibility with latest master changes

pbendersky committed:
Updated for compatibility with latest changes

pbendersky commented:
Here's the original comment I wrote on #252:

This PR enhances gitlab-plugin by adding two operation modes: **Modern** and **Legacy**.
**Legacy** is what we currently have, that works with older versions of GitLab.

**Modern** works with GitLab 8.1 or better, and provides an easier configuration of projects in Jenkins (a single Git repository is enough) and much better handling of Merge Requests (MR are fetched from origin and matched by SHA1 instead of comparing branch names).

This PR fixes the two issues I reported: #239 and #155.

omehegan commented:
@pbendersky thanks a lot for doing this work!

omehegan commented on src/main/java/com/dabsquared/gitlabjenkins/util/CommitStatusUpdater.java:
```diff
@@ -5,6 +5,7 @@
 import com.dabsquared.gitlabjenkins.gitlab.api.GitLabApi;
 import com.dabsquared.gitlabjenkins.gitlab.api.model.BuildState;
 import hudson.EnvVars;
+import hudson.model.AbstractBuild;
```
Are you using AbstractBuild here? Note that, in order to support Pipeline jobs, AbstractJob and AbstractBuild can't be used. See https://github.com/jenkinsci/pipeline-plugin/blob/master/DEVGUIDE.md

pbendersky replied:
I'm not. I'll add a commit removing it.

pbendersky committed:
Removed unused import of hudson.model.AbstractBuild

pbendersky committed:
Updated README with Modern / Legacy explanation

pbendersky committed:
Merge branch 'master' into merge-requests-1.2

pbendersky commented:
@coder-hugo @omehegan I'm keeping this PR updated with master, so it's easier to merge when you want to.
Please let me know if you want to do additional changes, or discuss any decisions I made.

pbendersky committed:
Merge remote-tracking branch 'upstream/master' into merge-requests-1.2

omehegan commented:
@pbendersky thanks! I'm going to try to test it in my environment tomorrow, where we use GitLab CE 7.14. That's probably the most I can do with it. @coder-hugo said that he will look through it as soon as he can.

pbendersky committed:
Merge branch 'master' into merge-requests-1.2

# Conflicts:
#	src/main/java/com/dabsquared/gitlabjenkins/cause/CauseData.java
#	src/main/java/com/dabsquared/gitlabjenkins/trigger/handler/merge/MergeRequestHookTriggerHandlerLegacyImpl.java

pbendersky committed:
Fixed issue with Commit Status Updater where it would crash if the build was initiated manually instead of through a GitLab trigger.

pbendersky committed:
Merge branch 'master' into merge-requests-1.2

omehegan commented:
@pbendersky I'm working on testing this. The only merge conflict is with changes to the README, so it should be easy to resolve that.

pbendersky commented:
Nice! Let me know if you need anything, or you have comments on the code.

omehegan commented:
@pbendersky there seems to be a bug in the existing HEAD of master, I'm getting an NPE when I trigger a build with that code. Unrelated to your change. I'll see if I can track it down.

tommyminds commented:
Is there any update on the status of this PR? Or on the next release (1.2.3) in general? There are some fixes in master that would be good to have released.

omehegan commented:
I determined yesterday that the current master has a bug that breaks compatibility with GitLab 7.14 and possibly everything before GitLab 8.5. I have a patch for it that just needs cleanup. I will merge that this week and then I can test this change.

omehegan commented:
OK, I have successfully tested this with GitLab 7.14, using the legacy mode, and it correctly triggers a build based on push or MR, and checks out the right branch. So that much looks good to me.

omehegan commented on src/main/java/com/dabsquared/gitlabjenkins/GitLabPluginMode.java:
```diff
@@ -0,0 +1,32 @@
+public enum GitLabPluginMode {
+    LEGACY,
+    MODERN;
+/*    
```
Do we need to keep all this commented out stuff?

omehegan commented on src/main/java/com/dabsquared/gitlabjenkins/GitLabPluginMode.java:
```diff
@@ -0,0 +1,32 @@
+/*
```
Maybe remove this generic license boilerplate.

omehegan commented on src/main/java/com/dabsquared/gitlabjenkins/cause/CauseData.java:
```diff
@@ -152,7 +152,7 @@ public String getMergeRequestDescription() {
         return mergeRequestId;
     }
-
+    
```
Whitespace?

omehegan commented on src/main/java/com/dabsquared/gitlabjenkins/trigger/handler/merge/AbstractMergeRequestHookTriggerHandler.java:
```diff
@@ -0,0 +1,47 @@
+/*
```
License boilerplate.

omehegan commented:
I posted some minor comments on the code. Also wanted to note that I tested with a Pipeline job and it works; this should imply that it will work with Freestyle jobs as well.

@coder-hugo pending comments from you before we merge this. I've reviewed as much as I think I can.

pbendersky committed:
- Merged master from upstream.
- Addressed comments in PR

pbendersky commented:
@omehegan Just fixed these minor issues, and merged from upstream, to make things easier.

omehegan commented on README.md:
```diff
@@ -21,7 +21,11 @@
     - [Parameterized builds](#parameterized-builds)
 - [Branch filtering](#branch-filtering)
 - [Build Tags](#build-tags)
+<<<<<<< HEAD
```
Looks like you committed the merge conflicts of README...

pbendersky replied:
Sorry, just added a new commit fixing it.

pbendersky committed:
Fixed readme

omorillo commented on README.md:
```diff
@@ -39,10 +50,17 @@ Version 1.2.0 of the plugin introduced improved logging for debugging purposes.
+The following versions of GitLab are supported by each of the modes:
+
+* Modern
+  * 8.1.x and newer
```
Why is version **8.1.x** listed in both **Modern** and **Legacy**? The modes should be mutually exclusive, shouldn't they?

pbendersky replied:
If you are using **>= 8.1** **Modern** is definitely better IMO. That said, nothing prevents you to use the **Legacy** mode. While I'd argue against these, I think there are some scenarios where you might want to leave **Legacy** mode:
1. Don't want to (or can't) mess with your Jenkins setup.
2. You want Jenkins to build pushes _on forks_ (**Modern** will build MR and commits added to MR, not commits to forks)

Basically, **Legacy** still works and **Modern** is opt-in (it's not even the default, although I think it has big advantages). If we want to push people to use **Modern**, we can change the wording here.

pbendersky committed:
Merged latest changes in master

omehegan commented:
@coder-hugo ping

coder-hugo commented on src/main/java/com/dabsquared/gitlabjenkins/trigger/handler/push/PushHookTriggerHandlerFactory.java:
```diff
@@ -12,20 +13,20 @@
             result.add(new PushHookTriggerHandlerImpl());
         }
-        if (triggerOpenMergeRequestOnPush == TriggerOpenMergeRequest.both) {
+        if (gitLabPluginMode == GitLabPluginMode.LEGACY && triggerOpenMergeRequestOnPush == TriggerOpenMergeRequest.both) {
```
With this you disable the functionality to rebuild open MRs if the target branch has changed for the modern mode. Why you are doing this?

pbendersky replied:
Because that case is handled by the Merge Request hook instead (if a MR is created and a commit added, the MR hook is called again).

coder-hugo replied:
Nope. I just tested this with GitLab 8.7:
1. I configured the job in Jenkins as you described in the README
2. I configured my GitLab repo to send MR events to the Jenkins
3. I created a MR for a test branch into the master -> a new build was triggerd
4. I pushed some changes to the master (target branch) -> no build was triggered as no event was sent by GitLab

pbendersky replied:
You are right. The case I was referring was new commits on the source branch (the user's fork). That gets handled by the MR trigger. I'll take a look at this, it might be a limitation of the "Modern" approach (if that's the case, I'll document it to reflect it).

pbendersky replied:
@coder-hugo just looked a bit at this... seems to be a limitation of the Modern mode. Since the user only has one repo configured, a new commit to `master` in the target branch can't be merged to the MR by Jenkins.
I suggest:
1. Updating the Readme to reflect this.
2. Ideally (I'll check how can it be done) change the drop down when selecting Modern mode, so the user can only select "Source branch" on when to re-build.
Thoughts?

pbendersky committed:
- Updated Readme to reflect that Modern Mode doesn't support re-building open MR when pushing to source branch.
- Changed the settings so when switching the plugin Mode the options for "triggerOpenMergeRequest" are refreshed and show only the possible values.

pbendersky committed:
Merge branch 'master' into merge-requests-1.2

pbendersky commented:
@coder-hugo just added a two commits:
- I clarified in the Readme that in Modern mode only pushes to the target repo will trigger a rebuild.
- I added a `target` value to `TriggerOpenMergeRequest.java`
- I changed `GitLabPushTrigger.java` so it reloads the options for `TriggerOpenMergeRequest` when you change the plugin mode and only shows the valid values for each mode.
  Hope this makes it clearer to users.

I also merged the latest changes from master, so it passes the checks here.

pbendersky committed:
Merge branch 'master' into merge-requests-1.2

coder-hugo commented:
@pbendersky it looks like you have interchanged the meaning of source and target branch of a MR. The source branch is the one that contains the changes that shall be merged to the target branch. So the modern mode supports re-building open MRs by pushes to the source but not to the target branch. And as the target branch is part of the configured repository it should be possible to re-build open MRs also for pushes to the target branch.

pbendersky commented:
Oh... I'll revert the source / target then.

So, a better approach (that would cover all the cases) would be to instead of fetching from `merge-requests/x` to fetch from `origin` and _merge_ the contents of `merge-requests/x`. That way, commits added to either source or target would get applied. Would that work? (of course, respecting the user settings for this).

pbendersky commented:
@coder-hugo I see one problem, however. Let's say there are 3 merge requests open, and a user commits to `master` in the upstream repository. How many builds should that trigger? If this works as you suggest, we would need to trigger four builds: one for the push to the upstream repo, and 3 for the updated merge requests. Is that correct?

pbendersky committed:
Changed target to source

omorillo commented:
@pbendersky I don't see any problem there. You would like to know if the master is still stable but also if the source branches can be safely merged to the target branch.

pbendersky commented:
@omorillo @coder-hugo I'm trying to wrap this up... Pending tasks (as discussed):
1. Merge origin with the merge request before build.
2. Add ability to trigger multiple builds on open merge requests when pushing to the target repo.

Now I need some help, and maybe you can give me a pointer... We currently create a `RevisionParameterAction` to tell Jenkins what revision to checkout. Is there any `Action` I can use to tell it to merge two repositories? I could not find it, and I'm not that familiar with Jenkins' API yet.

coder-hugo commented:
@pbendersky you don't need to add a special `Action` to the build for merging two branches. This will be configured as additional behavior (Merge before build) of the git SCM config. (see [here](https://github.com/jenkinsci/gitlab-plugin#jenkins-job-configuration))

pbendersky commented:
@coder-hugo right, but it loses some of the charm. As it is now, you simply set up `origin` on Jenkins and it will know how to fetch the merge requests from it. I'd like to keep that simplicity if possible. Maybe adding a second repository on build time? Other idea?

coder-hugo commented:
@pbendersky you can try to extend the `PreBuildMerge` of the git-plugin and pass the required `UserMergeOptions` for GitLab within the constructor to the super constructor. This extended class should then appear as a new additional behavior for the git SCM. This should keep the job setup as simple as possible.

pbendersky committed:
Merge branch 'master' into merge-requests-1.2

pbendersky committed:
fixed issues after merging

pbendersky committed:
Merge branch 'master' into merge-requests-1.2

# Conflicts:
#	README.md
#	src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java
#	src/main/java/com/dabsquared/gitlabjenkins/trigger/handler/merge/MergeRequestHookTriggerHandlerFactory.java
#	src/main/java/com/dabsquared/gitlabjenkins/trigger/handler/merge/MergeRequestHookTriggerHandlerLegacyImpl.java
#	src/main/java/com/dabsquared/gitlabjenkins/trigger/handler/push/PushHookTriggerHandlerFactory.java
#	src/main/java/com/dabsquared/gitlabjenkins/util/CommitStatusUpdater.java
#	src/test/java/com/dabsquared/gitlabjenkins/publisher/GitLabCommitStatusPublisherTest.java
#	src/test/java/com/dabsquared/gitlabjenkins/trigger/handler/merge/MergeRequestHookTriggerHandlerImplTest.java

pbendersky commented:
@coder-hugo @omehegan please take a look at #370. While I'm not giving up on this PR (I still think the setup with a single repository is easier to manage), the minor changes in configuration explained in #370 achieve the same result.

pabloa commented:
Hello, any news with this Pull Request? I am interested because #98

pbendersky commented:
I think it's a dead end at this point... With the configuration now in the README, I don't think this is needed. Did you try that?

omehegan commented:
Should we make a call on this one? It sounds like we should close it. Agreed?

pbendersky commented:
Agreed.
